### PR TITLE
chore: run only rust tests affected with changed files

### PR DESCRIPTION
**Pull request by MatejVukosav**
Tested with changed files
https://github.com/calimero-network/core/actions/runs/11698775125 

`./scripts/test.sh now accepts --all or --local {changed_files}`

Docs https://github.com/calimero-network/calimero-network.github.io/pull/48

### CODE REVIEW

**Review by chefsale — APPROVED** [review]
LGTM

File: scripts/test.sh [hunk 1]
```diff
@@ -5,13 +5,161 @@ set -ex
...
+            duplicate=true
+            break
+        fi
+    done
+    # If item is not a duplicate, add it to unique_array
+    if ! $duplicate; then
+        unique_changed_crates+=("$item")
+    fi
+done
+echo "Unique crates from changed files" "${unique_changed_crates[@]}"
```
Comment (miraclx) [inline scripts/test.sh:93]: `uniq` command?

File: scripts/test.sh [hunk 2]
```diff
@@ -5,13 +5,161 @@ set -ex
...
+    for dep in "${dependencies[@]}"; do
+        # Compare dependency with list of crates from workspace to skip external dependencies
+        calimero_dep_crate_path="${dep/calimero-/crates/}"
+        for crate in "${all_crates[@]}"; do
+            if [[ "$crate" == "$calimero_dep_crate_path" ]]; then
+                echo "Found matching dependency: $calimero_dep_crate_path"
+                dep_arr+=("$calimero_dep_crate_path")
+            fi
+        done
+    done
```
Comment (miraclx) [inline scripts/test.sh:117]: we shouldn't need to test dependencies of changed crates, but "dependents" of changed crates.

`calimero-node` depends on `calimero-context` which depends on `calimero-primitives`, if `calimero-context` changes we need to test that as well as `calimero-node`, but not necessarily `calimero-primitives`

File: scripts/test.sh [hunk 3]
```diff
@@ -5,13 +5,161 @@ set -ex
...
+
+# Run tests for each module and its dependencies
+echo "Running tests for affected modules and their dependencies..."
+echo "Crates to run tests:" "${crates_to_test[@]}"
+for crate in "${crates_to_test[@]}"; do
+    # Convert from folder name to package name
+    crate_package_name=("${crate/crates\//calimero-}")
+    if [[  "$crate_package_name" == "calimero-merod" ]]; then
+        echo "Testing crate merod"
+        cargo +nightly test -p "merod"
```
Comment (miraclx) [inline scripts/test.sh:157]: why test with nightly?

**Review by miraclx — COMMENTED** [review]
A lot of the crates don't compile on their own, for example `calimero-server`, `calimero-node-primitives`, `calimero-config` etc, relying instead for `merod` & `meroctl` to maximally define all the minimum feature requirements to satisfy all the crates.

Not an ideal situation, but quite the effort to untangle. It's tracked by #265 under

> 	- [ ] `cargo check` to ensure the `HEAD` builds right
>		- [ ] We need to walk the dep tree, creating a dependency graph and build each crate in order of dependence so we can guarantee all crates build independently as well as when linked. (See https://github.com/calimero-network/core/pull/258#discussion_r1593846781)
>		- [ ] Don't forget to build examples

This PR attempts to test each crate independently, it will break in cases where `cargo build` for a specific crate doesn't work.

if mode is not "all", this will not test apps and contracts? they don't have `calimero-` prefix, nor are they in `crates/*` folder.

We should revert this.

File: scripts/test.sh (hunk 2 above)
Comment (MatejVukosav) [reply scripts/test.sh:117]: It tests dependencies which imports changed crate

**Pull request merged**